Source/CarthageKit/Project.swift

```diff
@@ -1109,6 +1145,60 @@ public final class Project { // swiftlint:disable:this type_body_length
 					}
 			}
 	}
+    
+	/// Determines whether the requirements specified in this project's Cartfile.resolved
+	/// are compatible with the versions specified in the Cartfile for each of those projects.
+	///
+	/// Either emits a value to indicate success or an error.
+	public func validate(resolvedCartfile: ResolvedCartfile) -> SignalProducer<(), CarthageError> {
+		let resolvedCartfileProducer = SignalProducer(value: resolvedCartfile)
+			.promoteError(CarthageError.self)
+			.replayLazily(upTo: 1)
+
+		let dependencyRequirements = resolvedCartfileProducer
+			.flatMap(.merge) { (resolved: ResolvedCartfile) -> SignalProducer<[Dependency: [Dependency: VersionSpecifier]], CarthageError> in
+				return self.requirementsByDependency(resolvedCartfile: resolved, tryCheckoutDirectory: true)
+			}
+			.flatMap(.merge) { dependencyDict -> SignalProducer<[Dependency: [Dependency: VersionSpecifier]], CarthageError> in
+				// Invert the requirements dictionary: [A: [B: 1, C: 2]] -> [B: [A: 1], C: [A: 2]]
+				var dict: [Dependency: [Dependency: VersionSpecifier]] = [:]
+				for (dependency, requirements) in dependencyDict {
+					for (requiredDependency, requiredVersion) in requirements {
+						var requirements = dict[requiredDependency] ?? [:]
+
+						if requirements[dependency] != nil {
+							return SignalProducer(error: .duplicateDependencies([DuplicateDependency(dependency: dependency, locations: [])]))
+						}
+
+						requirements[dependency] = requiredVersion
+						dict[requiredDependency] = requirements
+					}
+				}
+				return SignalProducer(value: dict)
+			}
+
+		return resolvedCartfileProducer.map { (resolved: ResolvedCartfile) -> [Dependency: PinnedVersion] in
+				return resolved.dependencies
+			}
+			.combineLatest(with: dependencyRequirements)
+			.map { (dependencyInfo: ([Dependency: PinnedVersion], [Dependency: [Dependency: VersionSpecifier]])) -> [CompatibilityInfo] in
+				let (resolved, requirements) = dependencyInfo
+				var result: [CompatibilityInfo] = []
+				resolved.forEach { dependency, pinnedVersion in
+					// Skip non-semantic resolved versions and transitive dependencies without semantic versions
+					if case .success = SemanticVersion.from(pinnedVersion), let requirements = requirements[dependency] {
+						result.append(CompatibilityInfo(dependency: dependency, pinnedVersion: pinnedVersion, requirements: requirements))
+					}
+				}
+				return result
+			}
+			.flatMap(.merge) { (compatibilityInfos: [CompatibilityInfo]) -> SignalProducer<(), CarthageError> in
+				let incompatibleInfos: [CompatibilityInfo] = compatibilityInfos.filter { compatibilityInfo in
+					return !compatibilityInfo.incompatibleRequirements.isEmpty
+				}
+				return incompatibleInfos.isEmpty ? SignalProducer(value: ()) : SignalProducer(error: .invalidResolvedCartfile(incompatibleInfos))
+			}
```

Comment:
I think this would be better expressed if the `filter` was part of the signal chain.

You can do that by turning the `map` above this into a `flatMap`.

```swift
.flatMap(.concat) { dependencyInfo -> [CompatibilityInfo] in … }
.filter { !$0.incompatibleRequirements.isEmpty }
.collect()
.flatMap { incompatible in
  return incompatible.isEmpty ? SignalProducer(value: ()) : SignalProducer(error: …)
}
```